runtime/minor_gc.c

```diff
@@ -479,17 +479,24 @@ CAMLexport void caml_gc_dispatch (void)
 
 /* Called by [Alloc_small] when [caml_young_ptr] reaches [caml_young_limit].
    We have to either call memprof or the gc. */
-void caml_alloc_small_dispatch (tag_t tag, intnat wosize, int track)
+void caml_alloc_small_dispatch (tag_t tag, intnat wosize,
+                                enum caml_alloc_small_flags flags)
 {
   if (caml_young_ptr < caml_young_trigger){
+    int old_final_enabled = caml_final_enabled;
+    if(!(flags & CAML_FROM_CAML)) caml_final_enabled = 0;
     caml_young_ptr += Whsize_wosize (wosize);
     CAML_INSTR_INT ("force_minor/alloc_small@", 1);
     caml_gc_dispatch ();
     caml_young_ptr -= Whsize_wosize (wosize);
+    if(!(flags & CAML_FROM_CAML) && old_final_enabled) {
+      caml_final_enabled = 1;
+      caml_final_check_to_do();
+    }
```

Comment:
The presence of this dance to ensure that `caml_gc_dispatch` does not invoke finalisers suggests that `caml_gc_dispatch` has the wrong API. Perhaps `caml_gc_dispatch` should never invoke finalisers, and there should be a separate function that calls `caml_final_do_calls`, `caml_memprof_handle_postponed` and `caml_process_pending_signals`?

Also, `caml_final_check_to_do` looks like very similar logic to `caml_update_young_limit`, should these be the same function?

Reply:
> Also, caml_final_check_to_do looks like very similar logic to caml_update_young_limit, should these be the same function?

Hum, they seem really different to me. Perhaps you got confused with function names?

Reply:
No, I really meant it! Both of these functions check some global state, and if the global state indicates that async callbacks need to be run, they set `caml_young_limit` to `caml_young_alloc_end` (in native code mode). Of the three types of async callbacks (signals, memprof and finalisers), it's odd to me that the status of two of them is checked in `caml_update_young_limit` while the status of the third is checked here.

This API means that calling `caml_update_young_limit` is insufficient to correctly update the young limit: if finalisers may be pending, I must call `caml_final_check_to_do` as well.